### PR TITLE
DUPLO-20758: [TF] snapshot_window support in redis

### PR DESCRIPTION
**Pull request by pravin-duplocloud**
## Overview

 DUPLO-20758 snapshot_window support in redis
## Summary of changes

This PR does the following:

- snapshot_window support in redis

## Testing performed

- [ ] Using unit tests
- [X] Manually, on my local system
- [X] Manually, on a remote test system

## Describe any breaking changes

- None

### CODE REVIEW

File: duplocloud/resource_duplo_ecache_instance.go [hunk 1]
```diff
@@ -665,3 +675,42 @@ func removePatchVersion(version string) string {
...
+
+func isValidSnapshotWindow() schema.SchemaValidateFunc {
+	return func(i interface{}, k string) (warnings []string, errors []error) {
+		if i == nil {
+			return warnings, errors
+		}
+
+		v, ok := i.(string)
+		if !ok {
+			errors = append(errors, fmt.Errorf("expected type of %s to be string", k))
```
Comment (ddiaz-duplo) [inline duplocloud/resource_duplo_ecache_instance.go:687]:
```suggestion
			errors = append(errors, fmt.Errorf("the type of '%s' must be a string", k))
```

File: duplocloud/resource_duplo_ecache_instance.go [hunk 2]
```diff
@@ -665,3 +675,42 @@ func removePatchVersion(version string) string {
...
+			return warnings, errors
+		}
+
+		if v == "" {
+			return warnings, errors
+		}
+
+		times := strings.Split(v, "-")
+		if len(times) != 2 {
+			errors = append(errors, fmt.Errorf("expected %s in format '05:00-09:00'", k))
```
Comment (ddiaz-duplo) [inline duplocloud/resource_duplo_ecache_instance.go:697]:
```suggestion
			errors = append(errors, fmt.Errorf("the value of '%s' must follow the format 'HH:MM-HH:MM', for example '05:00-09:00'", k))
```

File: duplocloud/resource_duplo_ecache_instance.go [hunk 3]
```diff
@@ -665,3 +675,42 @@ func removePatchVersion(version string) string {
...
+			return warnings, errors
+		}
+
+		startTime, err1 := time.Parse("15:04", times[0])
+		endTime, err2 := time.Parse("15:04", times[1])
+		if err1 != nil || err2 != nil {
+			errors = append(errors, fmt.Errorf("expected %s in format '05:00-09:00'", k))
```
Comment (ddiaz-duplo) [inline duplocloud/resource_duplo_ecache_instance.go:704]:
```suggestion
			errors = append(errors, fmt.Errorf("the value of '%s' must be in the format 'HH:MM-HH:MM', for example '05:00-09:00'", k))
```

File: duplocloud/resource_duplo_ecache_instance.go [hunk 4]
```diff
@@ -665,3 +675,42 @@ func removePatchVersion(version string) string {
...
+			return warnings, errors
+		}
+
+		// Check if the input time is within the range
+		if !endTime.After(startTime) {
+			errors = append(errors, fmt.Errorf("expected %s of start-time should be smaller than end-time.", k))
```
Comment (ddiaz-duplo) [inline duplocloud/resource_duplo_ecache_instance.go:710]:
```suggestion
			errors = append(errors, fmt.Errorf("the start time in '%s' must be earlier than the end time", k))
```

**Review by ddiaz-duplo — COMMENTED** [review]
just adding some suggestions for improving the error messages

**Comment by zafarabbas** [comment]
Task linked: [DUPLO-20758 [TF] Ability to provide snapshot_window for automated backups](https://app.clickup.com/t/8655600/DUPLO-20758)

File: duplocloud/resource_duplo_ecache_instance.go (hunk 1 above)
Comment (pravin-duplocloud) [reply duplocloud/resource_duplo_ecache_instance.go]: done

File: duplocloud/resource_duplo_ecache_instance.go (hunk 2 above)
Comment (pravin-duplocloud) [reply duplocloud/resource_duplo_ecache_instance.go]: done

File: duplocloud/resource_duplo_ecache_instance.go (hunk 3 above)
Comment (pravin-duplocloud) [reply duplocloud/resource_duplo_ecache_instance.go]: done

File: duplocloud/resource_duplo_ecache_instance.go (hunk 4 above)
Comment (pravin-duplocloud) [reply duplocloud/resource_duplo_ecache_instance.go]: done

File: duplocloud/resource_duplo_ecache_instance.go [hunk 5]
```diff
@@ -186,6 +187,13 @@ func ecacheInstanceSchema() map[string]*schema.Schema {
 			Computed:     true,
 			ValidateFunc: validation.IntBetween(1, 35),
 		},
+		"snapshot_window": {
+			Description:      "Specify snapshot window limit The daily time range (in UTC) during which ElastiCache begins taking a daily snapshot of your node group (shard). Example: 05:00-09:00. If you do not specify this parameter, ElastiCache automatically chooses an appropriate time range.",
+			Type:             schema.TypeString,
```
Comment (Duplo-nikv) [inline duplocloud/resource_duplo_ecache_instance.go:192]: please mention if endtime is less than startime, endtime is considered to be of next day

**Review by Duplo-nikv — CHANGES_REQUESTED** [review]

File: duplocloud/resource_duplo_ecache_instance.go (hunk 5 above)
Comment (pravin-duplocloud) [reply duplocloud/resource_duplo_ecache_instance.go:192]: it was intentional ... after 24 hours in midnight
we need to take reminder.

Comment (pravin-duplocloud) [reply duplocloud/resource_duplo_ecache_instance.go:192]: the diff should be 30 min to 23.30 hour

Comment (pravin-duplocloud) [reply duplocloud/resource_duplo_ecache_instance.go:192]: Detail:   fmt.Sprintf("The time difference between the start ('%s') and end ('%s') must be at least 30 minutes. For example, '05:00-05:30' is valid, but '05:00-05:20' is not.", times[0], times[1]),

Comment (Duplo-nikv) [reply duplocloud/resource_duplo_ecache_instance.go:192]: I understand the logic, I just meant , add a note in document that if endtime is less than start time , the end time will be the difference carry forward to next day

**Review by Duplo-nikv — APPROVED** [review]
LGTM

File: duplocloud/resource_duplo_ecache_instance.go (hunk 5 above)
Comment (pravin-duplocloud) [reply duplocloud/resource_duplo_ecache_instance.go:192]: ok
its implied

**Review by Duplo-nikv — APPROVED** [review]
LGTM

**Pull request merged**